github/downloader_test.go

```diff
@@ -0,0 +1,153 @@
+package github
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/src-d/metadata-retrieval/testutils"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/oauth2"
+)
+
+// RepositoryTest struct to hold a test oracle for a repository
+type RepositoryTest struct {
+	Owner           string `json:"owner"`
+	Repository      string `json:"repository"`
+	Version         int    `json:"version"`
+	URL             string `json:"url"`
+	CreatedAt       string `json:"createdAt"`
+	IsPrivate       bool   `json:"isPrivate"`
+	IsArchived      bool   `json:"isArchived"`
+	HasWiki         bool   `json:"hasWiki"`
+	NumOfPRs        int    `json:"numOfPrs"`
+	NumOfPRComments int    `json:"numOfPrComments"`
+}
+
+// OrganizationTest struct to hold a test oracle for an organization
+type OrganizationTest struct {
+	Org               string `json:"org"`
+	Version           int    `json:"version"`
+	URL               string `json:"url"`
+	CreatedAt         string `json:"createdAt"`
+	PublicRepos       int    `json:"publicRepos"`
+	TotalPrivateRepos int    `json:"totalPrivateRepos"`
+	NumOfUsers        int    `json:"numOfUsers"`
+}
+
+// OnlineTests struct to hold the online tests
+type OnlineTests struct {
+	RepositoryTests    []RepositoryTest   `json:"repositoryTests"`
+	OrganizationsTests []OrganizationTest `json:"organizationTests"`
+}
+
+func checkToken(t *testing.T) {
+	if os.Getenv("GITHUB_TOKEN") == "" {
```

Comment:
From the PR desc, I'd say that the purpose of these tests, is to ensure that downloader can fetch an org and its repositories with its descendants, so having a `GITHUB_TOKEN` could be considered a hard requirement to pass the tests.

If so, I'd require such token instead of skipping the tests (which is not detected in Travis, and can only seen running the tests with `-v` flag.

```diff
 func checkToken(t *testing.T) {
-       if os.Getenv("GITHUB_TOKEN") == "" {
-               t.Skip("GITHUB_TOKEN is not set")
-               return
-       }
+       req := require.New(t)
+       req.NotEmpty(os.Getenv("GITHUB_TOKEN"), "GITHUB_TOKEN is not set")
 }
```

Reply:
Probably I misunderstood since I too agree that the tests should fail if there is no token. I got sidetracked because travis does not transfer env vars when you push an updated from a fork, so all PRs from a fork will fail. In any case I will change this.

Reply:
I didn't realize that detail, that will explain why integration tests requiring `GITHUB_TOKEN` are skipped in `sourced-ce` PR's (example: https://travis-ci.com/src-d/sourced-ce/jobs/240414245#L377), but they run on `master` (e.g. https://travis-ci.com/src-d/sourced-ce/jobs/240651163#L376).

And also why `ghsync` tests are using a custom `http.Client` [without authentication](https://github.com/src-d/ghsync/blob/master/utils/rate_test.go#L103), instead of the regular one, [with authentication](https://github.com/src-d/ghsync/blob/master/cmd/ghsync/subcmd/common.go#L132); approach that I think we can not take here because GH GraphQL API v4 requires authentication.

Considering that, I wonder if it'd be better if we replicate `source-ce` behavior about skipping the tests requiring `GITHUB_TOKEN` (in order to avoid having failed builds in PRs) and discuss an alternative in a new issue with the rest of apps and qa team.

Reply:
OK, I'll wait till later today and then I will skip the tests that require the token since I will another PR with tests that do not require it. 
So we can replicate the `source-ce` behavior where the master will run both online and offline tests, but in PRs forks will run only the offline ones based on http "recordings".

Reply:
@dpordomingo `GITHUB_TOKEN` in PRs is not only an apps team pain, I'm pretty sure that each team has bumped into it at least once and probably we should discuss it on the higher level than the team's one

As far as testing team has an editors access I suggest @kyrcha to create further PRs from origin branches

Anyways PRs from forks are still an open question

Reply:
I have one more PR ready to push from my fork and then I will start PRs from origin

Reply:
I wasn't aware of this behavior of env vars, but it totally makes sense.

 > I'm pretty sure that each team has bumped into it at least once and probably we should discuss it on the higher level than the team's one.

yes please! 👍